torch/optim/functional.py

```diff
@@ -0,0 +1,46 @@
+r"""Functional interface"""
+from torch import Tensor
+from typing import List
+
+# TODO: use foreach API in optim.functional to do all the computation
+
+def adagrad(params: List[Tensor],
+            grads: List[Tensor],
+            state_sums: List[Tensor],
+            step: int,
+            lr: float,
+            weight_decay: float,
+            lr_decay: float,
+            eps: float): 
+    r"""Functional API that performs Adagrad algorithm computation.
+
+    See :class:`~torch.optim.Adagrad` for details.
+    """
+
+    for (param, grad, state_sum) in zip(params, grads, state_sums):
+        if weight_decay != 0:
+            if grad.is_sparse:
+                raise RuntimeError("weight_decay option is not compatible with sparse gradients")
+            grad = grad.add(param, alpha=weight_decay)
+
+        clr = lr / (1 + (step - 1) * lr_decay)
+
+        if grad.is_sparse:
```

Comment:
note to self: I have seen users suggesting splitting out sparse/dense versions, see [comment](https://github.com/pytorch/pytorch/pull/30232#issuecomment-671203290). My current impression is that, given the scope of the current pull request, I would indeed prefer keeping both as done here instead of pushing the refactoring further..